File: decomp/ghidra/ExportFunctionTable.java
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;

import ghidra.app.script.GhidraScript;
import ghidra.program.model.address.Address;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionManager;
import ghidra.program.model.mem.Memory;
import ghidra.program.model.symbol.Symbol;
import ghidra.program.model.symbol.SymbolTable;

public class ExportFunctionTable extends GhidraScript {

	private static String jsonEscape(String value) {
		return value.replace("\\", "\\\\").replace("\"", "\\\"");
	}

	private static String hex32(long value) {
		return String.format("0x%08X", value & 0xffffffffL);
	}

	private static Long parseLongValue(String value) {
		return Long.decode(value);
	}

	@Override
	protected void run() throws Exception {
		String[] args = getScriptArgs();
		if (args.length < 3) {
			throw new IllegalArgumentException(
				"usage: ExportFunctionTable.java <output_json> <table_addr> <entry_count> [table_name] [pointer_size]");
		}

		if (currentProgram == null) {
			throw new IllegalStateException("No current program is open");
		}

		String outputJson = args[0];
		long tableAddrRaw = parseLongValue(args[1]);
		int entryCount = Integer.decode(args[2]);
		String tableName = args.length >= 4 ? args[3] : "function_table";
		int pointerSize = args.length >= 5 ? Integer.decode(args[4]) : 4;

		if (pointerSize != 4) {
			throw new IllegalArgumentException("Only 4-byte pointer tables are supported right now");
		}

		Memory memory = currentProgram.getMemory();
		FunctionManager functionManager = currentProgram.getFunctionManager();
		SymbolTable symbolTable = currentProgram.getSymbolTable();
		Address tableAddr = toAddr(tableAddrRaw);

		StringBuilder json = new StringBuilder();
		json.append("{\n");
		json.append("  \"program_name\": \"").append(jsonEscape(currentProgram.getName())).append("\",\n");
		json.append("  \"program_path\": \"")
			.append(jsonEscape(currentProgram.getDomainFile().getPathname()))
			.append("\",\n");
		json.append("  \"table_name\": \"").append(jsonEscape(tableName)).append("\",\n");
		json.append("  \"table_address\": \"").append(hex32(tableAddrRaw)).append("\",\n");
		json.append("  \"entry_count\": ").append(entryCount).append(",\n");
		json.append("  \"pointer_size\": ").append(pointerSize).append(",\n");
		json.append("  \"entries\": [\n");

		for (int index = 0; index < entryCount; index++) {
			Address entryAddr = tableAddr.add((long) index * pointerSize);
			long rawValue = memory.getInt(entryAddr) & 0xffffffffL;
			Address targetAddr = toAddr(rawValue);

			Function function = functionManager.getFunctionAt(targetAddr);
			if (function == null) {
				function = functionManager.getFunctionContaining(targetAddr);
			}

			String handlerName = null;
			String functionEntry = null;
			if (function != null) {
				handlerName = function.getName();
				functionEntry = function.getEntryPoint().toString();
			} else {
				Symbol symbol = symbolTable.getPrimarySymbol(targetAddr);
				if (symbol != null) {
					handlerName = symbol.getName();
					functionEntry = targetAddr.toString();
				}
			}

			json.append("    {\n");
			json.append("      \"index\": ").append(index).append(",\n");
			json.append("      \"opcode\": \"").append(String.format("0x%02X", index)).append("\",\n");
			json.append("      \"entry_address\": \"").append(entryAddr.toString()).append("\",\n");
			json.append("      \"handler_address\": \"").append(hex32(rawValue)).append("\",\n");
			if (handlerName == null) {
				json.append("      \"handler_name\": null,\n");
			} else {
				json.append("      \"handler_name\": \"").append(jsonEscape(handlerName)).append("\",\n");
			}
			if (functionEntry == null) {
				json.append("      \"function_entry\": null\n");
			} else {
				json.append("      \"function_entry\": \"").append(jsonEscape(functionEntry)).append("\"\n");
			}
			json.append(index == entryCount - 1 ? "    }\n" : "    },\n");
		}

		json.append("  ]\n");
		json.append("}\n");

		Path outputPath = Path.of(outputJson);
		Path parent = outputPath.getParent();
		if (parent != null) {
			Files.createDirectories(parent);
		}
		Files.write(outputPath, json.toString().getBytes(StandardCharsets.UTF_8));

		println("wrote " + entryCount + " entries to " + outputJson);
	}
}

File: decomp/ghidra/export_function_table.py
# Export a function pointer table from the current Ghidra program as JSON.
#
# Headless example:
#   analyzeHeadless <project_dir> <project_name> \
#     -import <binary> \
#     -scriptPath <script_dir> \
#     -postScript export_function_table.py <output_json> <table_addr> <entry_count> [table_name] [pointer_size]

import json
import os


def parse_int(value):
    return int(str(value), 0)


def to_hex(value):
    return "0x%08X" % (value & 0xFFFFFFFF)


def handler_name_for(symbol_table, function_manager, target_addr):
    function = function_manager.getFunctionAt(target_addr)
    if function is None:
        function = function_manager.getFunctionContaining(target_addr)
    if function is not None:
        return function.getName(), function.getEntryPoint().toString()

    symbol = symbol_table.getPrimarySymbol(target_addr)
    if symbol is not None:
        return symbol.getName(), target_addr.toString()

    return None, None


def main():
    args = getScriptArgs()
    if len(args) < 3:
        raise RuntimeError(
            "usage: export_function_table.py <output_json> <table_addr> <entry_count> [table_name] [pointer_size]"
        )

    output_json = args[0]
    table_addr_raw = parse_int(args[1])
    entry_count = parse_int(args[2])
    table_name = args[3] if len(args) >= 4 else "function_table"
    pointer_size = parse_int(args[4]) if len(args) >= 5 else 4

    if currentProgram is None:
        raise RuntimeError("No current program is open")

    memory = currentProgram.getMemory()
    function_manager = currentProgram.getFunctionManager()
    symbol_table = currentProgram.getSymbolTable()

    entries = []
    table_addr = toAddr(table_addr_raw)

    for index in range(entry_count):
        entry_addr = table_addr.add(index * pointer_size)
        if pointer_size != 4:
            raise RuntimeError("Only 4-byte pointer tables are supported right now")

        raw_value = memory.getInt(entry_addr) & 0xFFFFFFFF
        target_addr = toAddr(raw_value)
        handler_name, function_entry = handler_name_for(
            symbol_table, function_manager, target_addr
        )

        entries.append(
            {
                "index": index,
                "opcode": "0x%02X" % index,
                "entry_address": entry_addr.toString(),
                "handler_address": to_hex(raw_value),
                "handler_name": handler_name,
                "function_entry": function_entry,
            }
        )

    payload = {
        "program_name": currentProgram.getName(),
        "program_path": currentProgram.getDomainFile().getPathname(),
        "table_name": table_name,
        "table_address": to_hex(table_addr_raw),
        "entry_count": entry_count,
        "pointer_size": pointer_size,
        "entries": entries,
    }

    output_dir = os.path.dirname(output_json)
    if output_dir and not os.path.exists(output_dir):
        os.makedirs(output_dir)

    with open(output_json, "w") as handle:
        json.dump(payload, handle, indent=2, sort_keys=False)

    print("wrote %d entries to %s" % (entry_count, output_json))


main()
